distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,969 @@
+            self._default_start = 0
+            self._default_end = 0
+
+    def _message_completed(self):
+        """Push a completed message to the queue and reset per-message state"""
+        self._queue.put_nowait(self._frames)
+        self._nframes = None
+        self._frames = None
+        self._frame_lengths = None
+        self._frame_nbytes_remaining = 0
+
+    def connection_lost(self, exc=None):
+        self._transport = None
+        self._closed_waiter.set_result(None)
+
+        # Unblock read, if any
+        self._queue.put_nowait(_COMM_CLOSED)
+
+        # Unblock write, if any
+        if self._paused:
+            waiter = self._drain_waiter
+            if waiter is not None:
+                self._drain_waiter = None
+                if not waiter.done():
+                    waiter.set_exception(CommClosedError("Connection closed"))
+
+    def pause_writing(self):
+        self._paused = True
+
+    def resume_writing(self):
+        self._paused = False
+
+        waiter = self._drain_waiter
+        if waiter is not None:
+            self._drain_waiter = None
+            if not waiter.done():
+                waiter.set_result(None)
+
+    async def read(self) -> list[bytes]:
+        """Read a single message from the comm."""
+        # Even if comm is closed, we still yield all received data before
+        # erroring
+        if self._queue is not None:
+            out = await self._queue.get()
+            if out is not _COMM_CLOSED:
+                return out
+            self._queue = None
+        raise CommClosedError("Connection closed")
+
+    async def write(self, frames: list[bytes]) -> int:
+        """Write a message to the comm."""
+        if self.is_closed:
+            raise CommClosedError("Connection closed")
+        elif self._paused:
+            # Wait until there's room in the write buffer
+            drain_waiter = self._drain_waiter = self._loop.create_future()
+            await drain_waiter
+
+        # Ensure all memoryviews are in single-byte format
+        frames = [f.cast("B") if isinstance(f, memoryview) else f for f in frames]
+
+        nframes = len(frames)
+        frames_nbytes = [len(f) for f in frames]
+        # TODO: the old TCP comm included an extra `msg_nbytes` prefix that
+        # isn't really needed. We include it here for backwards compatibility,
+        # but this could be removed if we ever want to make another breaking
+        # change to the comms.
+        msg_nbytes = sum(frames_nbytes) + (nframes + 1) * 8
+        header = struct.pack(f"{nframes + 2}Q", msg_nbytes, nframes, *frames_nbytes)
+
+        if msg_nbytes < 4 * 1024:
+            # Always concatenate small messages
+            buffers = [b"".join([header, *frames])]
+        else:
+            buffers = coalesce_buffers([header, *frames])
+
+        if len(buffers) > 1:
+            self._transport.writelines(buffers)
+        else:
+            self._transport.write(buffers[0])
```

Comment:
Thoughts on just doing `writelines` always?

Reply:
I've commented on this elsewhere above. Some `transport` implementations `writelines` is less efficient than `write`, and it's cheap to do this check here ourselves.